src/channel_manager.ts

```diff
@@ -0,0 +1,527 @@
+  [key in ChannelManagerEventTypes]: ChannelManagerEventHandlerNames;
+} = {
+  'channel.deleted': 'channelDeletedHandler',
+  'channel.hidden': 'channelHiddenHandler',
+  'channel.truncated': 'channelTruncatedHandler',
+  'channel.visible': 'channelVisibleHandler',
+  'message.new': 'newMessageHandler',
+  'member.updated': 'memberUpdatedHandler',
+  'notification.added_to_channel': 'notificationAddedToChannelHandler',
+  'notification.message_new': 'notificationNewMessageHandler',
+  'notification.removed_from_channel': 'notificationRemovedFromChannelHandler',
+};
+
+export type ChannelManagerOptions = {
+  allowNewMessagesFromUnfilteredChannels?: boolean;
+  lockChannelOrder?: boolean;
+};
+
+export const DEFAULT_CHANNEL_MANAGER_OPTIONS = {
+  allowNewMessagesFromUnfilteredChannels: true,
+  lockChannelOrder: false,
+};
+
+export class ChannelManager<SCG extends ExtendableGenerics = DefaultGenerics> {
+  public readonly state: StateStore<ChannelManagerState<SCG>>;
+  private client: StreamChat<SCG>;
+  private unsubscribeFunctions: Set<() => void> = new Set();
+  private eventHandlers: Map<string, EventHandlerType<SCG>> = new Map();
+  private eventHandlerOverrides: Map<string, EventHandlerOverrideType<SCG>> = new Map();
+  private options: ChannelManagerOptions = {};
+  private stateOptions: ChannelStateOptions = {};
+
+  constructor({
+    client,
+    eventHandlerOverrides = {},
+    options = {},
+  }: {
+    client: StreamChat<SCG>;
+    eventHandlerOverrides?: ChannelManagerEventHandlerOverrides<SCG>;
+    options?: ChannelManagerOptions;
+  }) {
+    this.client = client;
+    this.state = new StateStore<ChannelManagerState<SCG>>({
+      channels: [],
+      pagination: {
+        isLoading: false,
+        isLoadingNext: false,
+        hasNext: false,
+        filters: {},
+        sort: {},
+        options: { limit: 10, offset: 0 },
+      },
+      ready: false,
+    });
+    this.setEventHandlerOverrides(eventHandlerOverrides);
+    this.setOptions(options);
+    this.eventHandlers = new Map(
+      Object.entries({
+        channelDeletedHandler: this.channelDeletedHandler,
+        channelHiddenHandler: this.channelHiddenHandler,
+        channelVisibleHandler: this.channelVisibleHandler,
+        memberUpdatedHandler: this.memberUpdatedHandler,
+        newMessageHandler: this.newMessageHandler,
+        notificationAddedToChannelHandler: this.notificationAddedToChannelHandler,
+        notificationNewMessageHandler: this.notificationNewMessageHandler,
+        notificationRemovedFromChannelHandler: this.notificationRemovedFromChannelHandler,
+      }) as [string, EventHandlerType<SCG>][],
+    );
+  }
+
+  public setChannels = (valueOrFactory: ChannelSetterParameterType<SCG>) => {
+    this.state.next((current) => {
+      const { channels: prevChannels } = current;
+      let newValue = valueOrFactory;
+      if (newValue && typeof newValue === 'function') {
+        newValue = newValue(prevChannels);
+      }
```

Comment:
Let's reuse `StateStore.isPatch` for this part (needs to be exported from `store.ts`), also let's change the naming as it's not a previous state yet but state before it gets commited to handlers (which it does not have to).

```suggestion
      const { channels: currentChannels } = current;
      const newChannels = isPatch(valueOrFactory) ? valueOrFactor(currentChannels) : valueOrFactory;
```

Reply:
Can do, thanks !